### PR TITLE
Bugfix/#173 protect geometry items

### LINKED ISSUE #173: GameEngine provides original hints
Searcher.move gets the original object and is able to manipulate it.
Fix in:
* [ ] Treasure
* [ ] SearchPath
* [ ] Hints

### LINKED ISSUE #186: Invariant: SearchPath 0 or ≥2 Points
SearchPath need only to have 0 or ≥2 Points at each point of time, since adding one points causes
either a point in the game or complicated case differentation, since LineSegments
must not contain one point.

Fixes:
* [x] On one-elemented SearchPaths, there is only a point on the screen. This should get removed, since only the google-maps-pin should show us the player location.
* [ ] Remove SearchPathPrototypes. (Optional)
* [x] Restrict the SearchPaths on only containing 0 or ≥2 points.
* [x] SearchPath.getFirstPoint() & SearPatch.getLastPoint() may return null, which causes an error in the statistics! Fix this.
* [ ] Delete `selected` variable from `Selectable`.

### PR DESCRIPTION
**Pull request by DorianRec**
This contains content of PR's #95 and #185, merge these first.

### Changes:
* All changes of #185 
* Added a class **SeachPathPrototype**, which the searchers build in order to give this to the GameEngine. The GameEngine converts this into a SearchPath, which is still used by all other instances (except the Searchers), like the way before.
* Changes Point variables to Coordinate variables on many places, since the Coordinates are more light weight and save more runtime and space.
* Made the `SearchPath` contain either 0 point or ≥2 points, like the way, the LineStrings to. Thus, it is easier to create a LineString without case differentations.
* Made the `searchPath` know the `searcherStartCoordinate` and `searcherEndCoordinate`. Thus, we do not need to request first or last point of a potential empty search path.
* Made the Selectables (e.g. Hint, SearchPath, Treasure) secure from changing by removing all setters. Getters will return copys of the Geometry objects or clones of GeometryItems. Also, the searchers will build up SearchPathPrototypes objects instead of SearchPath objects, which they can modify like they want. Later, the GameEngine converts this to a SearchPath which is secure from getting modified.
* Outsourced the method `static boolean located(List<Coordinate> coordinates, Coordinate treasure)` into the GameEngine, as it was before, since the GameEngine has to decide about the game-rules instead of the SearchPaths do.

Fixes #173
Fixes #186

### CODE REVIEW

File: src/main/java/com/treasure/hunt/game/GameManager.java [hunk 1]
```diff
@@ -384,4 +398,75 @@ public GameManager copy(Kryo kryo) {
...
     }
+
+    /**
+     * @param coordinate the point on the canvas, we want to get the closest {@link GeometryType} to.
+     * @param distance   the maximum distance to a potential {@link GeometryItem}.
+     * @return a sorted list, containing the nearest {@link GeometryItem}'s to {@code coordinate}, with a maximum distance of {@code distance}.
+     */
+    private List<Selectable> pickGeometryItem(Coordinate coordinate, double distance) {
+        List<Turn> visibleTurns = getVisibleTurns();
+        List<Selectable> selectables = new ArrayList<>();
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/game/GameManager.java:404]: Why is this here and not in s separate Class?

File: src/main/java/com/treasure/hunt/game/GameManager.java [hunk 2]
```diff
@@ -384,4 +398,75 @@ public GameManager copy(Kryo kryo) {
...
+        if (visibleTurns.size() < 1) {
+            return new ArrayList<>();
+        }
+
+        Point mouse = JTSUtils.GEOMETRY_FACTORY.createPoint(coordinate);
+
+        for (Turn t : visibleTurns) {
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/game/GameManager.java]: t is the same as a,b,c

File: src/main/java/com/treasure/hunt/game/GameManager.java [hunk 3]
```diff
@@ -51,6 +53,10 @@
 @Slf4j
 public class GameManager implements KryoSerializable, KryoCopyable<GameManager> {
 
+    public static final double MOUSE_RECOGNIZE_DISTANCE = 0.2;
+    private List<Selectable> selectables = new ArrayList<>();
+    private int selectablesIndex = 0;
+    private Coordinate lastMouseClick;
     /**
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/game/GameManager.java:60]: Why is that here?

File: src/main/java/com/treasure/hunt/game/Turn.java [hunk 4]
```diff
@@ -28,29 +29,72 @@
...
     /**
      * The current location of the treasure.
      */
-    private Point treasureLocation;
+    private Treasure treasure;
+
+    public Turn(Hint hint, SearchPath searchPath, Coordinate treasure) {
+        this.hint = hint;
+        this.searchPath = searchPath;
+        this.treasure = new Treasure(treasure);
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/game/Turn.java:46]: Is that your way of trying to deep clone stuff?

File: src/main/java/com/treasure/hunt/game/Turn.java [hunk 5]
```diff
@@ -28,29 +29,72 @@
...
         if (searchPath != null) {
-            output.addAll(getSearchPathGeometries(movementStart));
+            output.addAll(getSearchPathGeometries());
+        }
+
+        if (treasure != null) {
+            output.addAll(getTreasureGeometries());
+        }
+
+        output.sort(Comparator.comparingInt(a -> a.getGeometryStyle().getZIndex()));
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/game/Turn.java]: a!

File: src/main/java/com/treasure/hunt/game/Turn.java [hunk 6]
```diff
@@ -28,29 +29,72 @@
...
+        }
+
+        if (searchPath != null) {
+            searchPath.setSelected(false);
+        }
+
+        if (treasure != null) {
+            treasure.setSelected(false);
         }
+    }
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/game/Turn.java]: What has this to do here

File: src/main/java/com/treasure/hunt/jts/other/ImageItem.java [hunk 7]
```diff
@@ -25,6 +27,7 @@ public ImageItem(Coordinate position, int maxWidth, int maxHeight, String imageP
         this.maxWidth = maxWidth;
         this.maxHeight = maxHeight;
         this.imagePath = imagePath;
+
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/jts/other/ImageItem.java]: remove

File: src/main/java/com/treasure/hunt/view/CanvasController.java [hunk 8]
```diff
@@ -48,19 +58,58 @@ public void makeCanvasResizable() {
...
             return;
         }
+
+        if (dragged) {
+            dragged = false;
+            return;
+        }
+        dragged = false;
+
+        Coordinate c = transformation.revert(dragStart.toCoordinate());
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/view/CanvasController.java]: c

File: src/main/java/com/treasure/hunt/view/CanvasController.java [hunk 9]
```diff
@@ -39,6 +47,8 @@ public void initialize() {
         renderer.addAdditional("grid", new GeometryItem<>(new Grid(), GeometryType.GRID));
         transformation.getScaleProperty().addListener(invalidation -> drawShapes());
         transformation.getOffsetProperty().addListener(invalidation -> drawShapes());
+
+        //subscribeToGeometryItem();
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/view/CanvasController.java]: commented code

**Review by 1-alex98 — CHANGES_REQUESTED** [review]

**Comment by 1-alex98** [comment]
AHHH one change-> one issue-> one PR 

You know what happens if I/we/whoever dislikes one of the changes... You will sit there redoing them or trying to separate them

**Comment by 1-alex98** [comment]
Also it gets a horror to review, I don't freaking know what belongs to what and then there are 50 Files changed at once

File: src/main/java/com/treasure/hunt/game/Turn.java (hunk 4 above)
Comment (DorianRec) [reply src/main/java/com/treasure/hunt/game/Turn.java:46]: No.

File: src/main/java/com/treasure/hunt/game/Turn.java (hunk 5 above)
Comment (DorianRec) [reply src/main/java/com/treasure/hunt/game/Turn.java]: I do not wrote this line, but ill fix it

File: src/main/java/com/treasure/hunt/game/Turn.java (hunk 6 above)
Comment (DorianRec) [reply src/main/java/com/treasure/hunt/game/Turn.java]: outdated, since selected field will be removed

File: src/main/java/com/treasure/hunt/game/GameManager.java (hunk 1 above)
Comment (DorianRec) [reply src/main/java/com/treasure/hunt/game/GameManager.java]: > AHHH one change-> one issue-> one PR
> 
> You know what happens if I/we/whoever dislikes one of the changes... You will sit there redoing them or trying to separate them

No problem, when the PR's are solved in the chronological order, which is important in this case. Had to fix things in the older PR's which are important to solve Problems in the newer PR's.
